### PR TITLE
[ParseSQL] Support queries with {[variable}}s and {{field filter}}s in them

### LINKED ISSUE #39953: Support parsing native queries with a field filter in them
Part of [ParseSQL](https://github.com/metabase/metabase/issues/36911)

Current idea is to replace the field filter with a gensym on the Metabase side, let Macaw parse it as normal, then replace the sentinel value with the field filter when it comes back to Metabase.

### LINKED ISSUE #39954: Support parsing native queries with variables in them
Part of [ParseSQL](https://github.com/metabase/metabase/issues/36911)

Current idea is to replace the variable with a gensym on the Metabase side, let Macaw parse it as normal, then replace the sentinel value with the field filter when it comes back to Metabase.

### LINKED ISSUE #41242: Support parsing native queries with a {{#card-id}} ref in them

### PR DESCRIPTION
**Pull request by tsmacdonald**
[Fixes #39954]

[Fixes #39953]

[Fixes #41242]

### CODE REVIEW

**Comment by piranha** [comment]
This looks good, but I can't help but wonder if this logic should be unified with how parameter substitution is done in query processor? Or is it different enough that there is no sense?

**Comment by tsmacdonald** [comment]
@piranha I did get a blessing from the QPD team: see the thread here https://metaboat.slack.com/archives/C04DN5VRQM6/p1712563635335469

File: test/metabase/models/query_field_test.clj [hunk 1]
```diff
@@ -68,9 +70,22 @@
...
+      (testing "Columns referenced via field filters are still found"
+        (trigger-parse! card-id
+                        (mt/native-query {:query "SELECT tax FROM orders WHERE {{adequate_total}}"
+                                          :template-tags {"adequate_total"
+                                                          {:type         :dimension
+                                                           :name         "adequate_total"
+                                                           :display-name "Total is big enough"
+                                                           :dimension    [:field (mt/id :orders :total)
+                                                                          {:base-type :type/Number}]
+                                                           :widget-type  :number/>=}}}))
```
Comment (tsmacdonald) [inline test/metabase/models/query_field_test.clj:87]: it seemed excessive to have tests here for every type of `{{thing}}`. I figure the `parameter-substitution-test` is for all the variations of `{{stuff}}` and this test is just to make sure the integration works. Unit testing!

File: test/metabase/native_query_analyzer/parameter_substitution_test.clj [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                    :location/state
+                    :location/zip_code
+                    :location/country}
+        expected  (set/difference (into #{} (keys lib.schema.parameter/types))
+                                  blocklist)
+        tag->type (fn [{:keys [type widget-type]}] (if (= type :dimension) widget-type type))]
+    (is (set/subset? expected (into #{} (keys nqa.sub/default-values)))
+        "Ensure that you have a default value for all types defined in lib.schema.parameter/types")
+    (is (set/subset? expected (into #{} (map tag->type (vals @tag-definitions))))
+        "Ensure that you have a test for each type defined in lib.schema.parameter/types")))
```
Comment (tsmacdonald) [inline test/metabase/native_query_analyzer/parameter_substitution_test.clj:216]: this should save us if new field filter types are ever added

**Review by sansolovyov — APPROVED** [review]
I'm kind of blind I guess, did not see it's calling `qp.compile/compile`. Seems good then!

**Pull request merged**